Herndlhofer lab, Powershell 5.1

NikitHern opened the pull request:
Инструкция по запуску:
Запустить Windows Powershell ISE (В обычной powershell консоли просматривать код файла неудобно, и подсказок он не дает)
Открыть файл FileHashAsync.ps1
Запустить его
Вызвать в консоли снизу функцию HashAsync с параметрами -FilePath,  -BlockSize,  -isAsync,  -Algorithm (Я могу сделать, чтобы сразу вызывалась функция, по запуску файла, но тогда не будут предложены параметры)
при вызове двух последних параметров будут подсказки с поддерживаемыми параметрами на вход

NikitaHerndlhofer committed:
Herndlhofer lab ready

NikitaHerndlhofer committed:
Added timer

NikitHern committed:
Fixed max bytes count

Andrew414 commented on Herndlhofer/FileHashAsync.ps1:
```diff
@@ -0,0 +1,109 @@
+  }
+  return $chunkNum
+}
+function HashAsync
```
The main idea of this task is to make it script-ready so it can be used in processing gigabytes of data without user interaction. I didn't find a way to run it as script from powershell.

NikitHern replied:
I didn't knew how do i have to make it work, so i said in the description, that i can make it work both ways. "(Я могу сделать, чтобы сразу вызывалась функция, по запуску файла, но тогда не будут предложены параметры)"

Andrew414 commented on Herndlhofer/FileHashAsync.ps1:
```diff
@@ -0,0 +1,109 @@
+    [Parameter(Mandatory)]
+    [int]$BlockSize,
+    [Parameter(Mandatory = $true)]
+    [ValidateSet('SHA1', 'SHA256', 'SHA384', 'SHA512', 'MD5', 'MACTripleDES', 'RIPEMD160')]
```
No CRC32, a lot of excess algorithms instead.

Andrew414 commented on Herndlhofer/FileHashAsync.ps1:
```diff
@@ -0,0 +1,109 @@
+    [Parameter(Mandatory = $true)]
+    [ValidateSet('SHA1', 'SHA256', 'SHA384', 'SHA512', 'MD5', 'MACTripleDES', 'RIPEMD160')]
+    [String]$Algorithm,
+    [Parameter(Mandatory)]
```
This parameter is not mandatory actually, the task should be working in async mode if nothing is passed.

Andrew414 commented on Herndlhofer/FileHashAsync.ps1:
```diff
@@ -0,0 +1,109 @@
+        throw [System.IO.Exception] 'blocksize is not power of 2.'
+        return
+      }
+      if (!($BlockSize -lt 1048576 -or $BlockSize -gt 1024))
```
Should be 
`if ($BlockSize -gt 1048576 -or $BlockSize -lt 1024)`
Currently it accepts block 512, for example.

Andrew414 commented on Herndlhofer/FileHashAsync.ps1:
```diff
@@ -0,0 +1,109 @@
+        }
+        $RunspacePool.Close()
+        $RunspacePool.Dispose()
+        <# Method with jobs
```
Not sure why such a big comment is left. I think it's not needed.

Andrew414 reviewed (approved):
Hi, 

I'm sorry for the very delayed answer. Anyway, you have a lot of time to fix something in the task (it was submitted in the first possible day and it took 3 weeks for me to give you my feedback so your deadline moves to April 11th).

Also, if you want, I can rewrite all my comments in russian. If it's more comfortable for you.

The score for this task is 6/10. Here is why:
- 3/4 for main task. There are several bugs - wrong block sizes are accepted (in fact, any number that is power of 2, like 64, 512 or 2^30). Also, the script can't be run as an actual script - by silently passing its parameters without user interaction. Or I couldn't find any way to do this. See my comments for more info.
- 3/3 for being in time. You finished the task really fast.
- 0/1 for GUI, obviously.
- 0/1 for additional algorithms. CRC32 is mandatory for this point but it wasn't implemented.
- 0/1 for performance measurement. As was said in the task description, the following 4 values should be printed: 1) Time spent on **reading**, 2) Speed on **reading**, 3) Time spend on **hash calculation**, 4) Speed of **hash calculation**. You simply measured the whole time of processing one file, which is not enough.

So, you can fix my comments within 4 weeks or you can do nothing, I already put your score to the table.

Thank you,
Andrey Fedchuk

NikitHern committed:
Made changes

1. Script runs silently
2. IsAsync not mandatory
3. Fixed blocksize check
4. Deleted redundant comment
